File: reference/src/database/crud.py
from datetime import datetime, timezone
from uuid import UUID, uuid4
from reference.src.database.connection import Cursor, db_connection
from reference.src.enums.command_enums import CommandStatus, SessionStatus
from reference.src.resources.commands import DatabaseCommand, MainCommand


@db_connection
def get_next_session(cur: Cursor) -> datetime | None:
    """
    Returns the start_time of the next upcoming session — the soonest
    PENDING session with a start_time still in the future. Returns None
    if no such session exists.
    """
    cur.execute(
        """
        SELECT start_time
        FROM transactional.sessions
        WHERE status = %s AND start_time > now()
        ORDER BY start_time ASC
        LIMIT 1
        """,
        (SessionStatus.PENDING.value,),
    )
    row = cur.fetchone()
    return row[0] if row is not None else None


@db_connection
def get_next_session_id(cur: Cursor) -> UUID | None:
    """
    Returns the ID of the next upcoming pending session.
    """
    cur.execute(
        """
        SELECT id
        FROM transactional.sessions
        WHERE status = %s
          AND start_time > NOW()
        ORDER BY start_time ASC
        LIMIT 1
        """,
        (SessionStatus.PENDING.value,),
    )

    row = cur.fetchone()
    return row[0] if row is not None else None


# TODO: add error handling if start_time is None
@db_connection
def update_current_session_status(
    cur: Cursor,
    status: SessionStatus,
    start_time: datetime | None,
) -> None:
    cur.execute(
        """
        UPDATE transactional.sessions
        SET status = %s
        WHERE id = (
            SELECT id
            FROM transactional.sessions
            WHERE start_time = %s
            ORDER BY start_time ASC
            LIMIT 1
        )
        """,
        (status.value, start_time),
    )


@db_connection
def get_main_command_by_id(cur: Cursor, command_id: int | None) -> MainCommand | None:
    """
    Replacement for MainCommandWrapper().get_by_id(command.type_).
    Fetches a single row from main.commands.
    """
    if command_id is None:
        return None
    cur.execute(
        """
        SELECT id, name, params, format, data_size, total_size, priority
        FROM main.commands
        WHERE id = %s
        """,
        (command_id,),
    )
    row = cur.fetchone()
    if not row:
        return None
    return MainCommand.from_row(row)


@db_connection
def get_all_commands_by_status(
    cur: Cursor, status: CommandStatus
) -> list[DatabaseCommand]:
    """
    Replacement for CommandsWrapper().get_all_by(status=CommandStatus.PENDING).
    Fetches all rows from transactional.commands matching a given status.
    """
    cur.execute(
        """
        SELECT id, user_id, status, type_, params, created_at, packet_id, sequence_index
        FROM transactional.commands
        WHERE status = %s
        """,
        (status.value,),
    )
    rows = cur.fetchall()
    return [DatabaseCommand.from_row(row) for row in rows]


@db_connection
def get_all_commands_next_session(
    cur: Cursor,
    status: CommandStatus,
) -> list[DatabaseCommand]:
    session_id = get_next_session_id()

    if session_id is None:
        return []

    cur.execute(
        """
        SELECT
            id,
            user_id,
            status,
            type_,
            params,
            created_at,
            packet_id,
            sequence_index,
            session_id
        FROM transactional.commands
        WHERE session_id = %s
          AND status = %s
        """,
        (
            session_id,
            status.value,
        ),
    )

    return [DatabaseCommand.from_row(row) for row in cur.fetchall()]


@db_connection
def update_command_status(cur: Cursor, command_id: UUID, status: CommandStatus) -> None:
    """
    Replacement for CommandsWrapper().update(command_id, {"status": ...}).
    Used in build_queue() (-> SCHEDULED), queue_to_packet() (-> ONGOING),
    and clear_queue() (-> COMPLETED).
    """
    cur.execute(
        """
        UPDATE transactional.commands
        SET status = %s
        WHERE id = %s
        """,
        (status.value, command_id),
    )


@db_connection
def update_command_response(cur: Cursor, command_id: UUID, response: str) -> None:
    cur.execute(
        """
    UPDATE transactional.commands
    SET response = %s
    WHERE id = %s
    """,
        (response, command_id),
    )


@db_connection
def create_telemetry(
    cur: Cursor,
    type_: int,
    value: object,
    timestamp: datetime,
) -> None:
    """
    Insert one telemetry value into transactional.telemetry.

    timestamp must be timezone-aware. It is normalized to UTC before insertion.
    """

    if timestamp.tzinfo is None:
        raise ValueError("timestamp must be timezone-aware")

    timestamp_utc = timestamp.astimezone(timezone.utc)

    cur.execute(
        """
        INSERT INTO transactional.telemetry (
            id,
            type_,
            value,
            timestamp
        )
        VALUES (%s, %s, %s, %s);
        """,
        (
            str(uuid4()),
            type_,
            str(value),
            timestamp_utc,
        ),
    )


@db_connection
def clear_telemetry_by_type(cur: Cursor, telemetry_type: int) -> int:
    """
    Deletes every telemetry row whose type_ matches telemetry_type.

    Returns the number of rows deleted.
    """
    cur.execute(
        """
        DELETE FROM transactional.telemetry
        WHERE type_ = %s
        """,
        (telemetry_type,),
    )

    deleted_count = cur.rowcount

    return deleted_count

File: src/database/crud.go
package database

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"

	"state-machine/src/database/gen"
	"state-machine/src/enums"
)

// ErrInvalidRow is returned when a row fails a data-integrity check that would
// normally be enforced by the ORM/DB layer.
var ErrInvalidRow = errors.New("invalid row")

// GetNextSession returns the start time of the next upcoming session — the
// soonest PENDING session with a start_time still in the future. Returns nil
// if no such session exists.
func GetNextSession(ctx context.Context) (*time.Time, error) {
	startTime, err := Queries().GetNextSession(ctx, gen.SessionstatusPENDING)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get next session: %w", err)
	}
	return &startTime, nil
}

// GetNextSessionID returns the ID of the next upcoming pending session, or nil
// if no such session exists.
func GetNextSessionID(ctx context.Context) (*uuid.UUID, error) {
	id, err := Queries().GetNextSessionID(ctx, gen.SessionstatusPENDING)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get next session id: %w", err)
	}
	return &id, nil
}

// UpdateCurrentSessionStatus sets the status of the session starting at
// startTime.
func UpdateCurrentSessionStatus(
	ctx context.Context,
	status enums.SessionStatus,
	startTime time.Time,
) error {
	sessionStatus, err := sessionStatus(status)
	if err != nil {
		return err
	}
	if startTime.IsZero() {
		return fmt.Errorf("failed to update session status: %w: start time is unset", ErrInvalidRow)
	}

	err = Queries().UpdateCurrentSessionStatus(ctx, gen.UpdateCurrentSessionStatusParams{
		Status:    sessionStatus,
		StartTime: startTime,
	})
	if err != nil {
		return fmt.Errorf("failed to update session status: %w", err)
	}
	return nil
}

// GetMainCommandByID fetches a single row from main.commands. Returns nil if no
// command with that ID exists.
func GetMainCommandByID(ctx context.Context, commandID int32) (*gen.MainCommand, error) {
	command, err := Queries().GetMainCommandByID(ctx, commandID)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get main command %d: %w", commandID, err)
	}
	if err := validateMainCommand(command); err != nil {
		return nil, fmt.Errorf("main command %d: %w", commandID, err)
	}
	return &command, nil
}

// GetAllCommandsByStatus fetches all rows from transactional.commands matching a
// given status.
func GetAllCommandsByStatus(
	ctx context.Context,
	status enums.CommandStatus,
) ([]gen.TransactionalCommand, error) {
	commandStatus, err := commandStatus(status)
	if err != nil {
		return nil, err
	}

	commands, err := Queries().GetAllCommandsByStatus(ctx, commandStatus)
	if err != nil {
		return nil, fmt.Errorf("failed to get %v commands: %w", status, err)
	}
	return commands, nil
}

// GetAllCommandsNextSession fetches the commands with the given status that
// belong to the next upcoming pending session. Returns an empty slice if there
// is no such session.
func GetAllCommandsNextSession(
	ctx context.Context,
	status enums.CommandStatus,
) ([]gen.TransactionalCommand, error) {
	commandStatus, err := commandStatus(status)
	if err != nil {
		return nil, err
	}

	commands, err := Queries().GetAllCommandsNextSession(ctx, gen.GetAllCommandsNextSessionParams{
		Status:   gen.SessionstatusPENDING,
		Status_2: commandStatus,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to get %v commands for next session: %w", status, err)
	}
	return commands, nil
}

// UpdateCommandStatus moves a command to a new status. Used in BuildQueue
// (-> SCHEDULED), QueueToPacket (-> ONGOING) and ClearQueue (-> COMPLETED).
func UpdateCommandStatus(ctx context.Context, commandID uuid.UUID, status enums.CommandStatus) error {
	commandStatus, err := commandStatus(status)
	if err != nil {
		return err
	}

	err = Queries().UpdateCommandStatus(ctx, gen.UpdateCommandStatusParams{
		Status: commandStatus,
		ID:     commandID,
	})
	if err != nil {
		return fmt.Errorf("failed to update status of command %v: %w", commandID, err)
	}
	return nil
}

// UpdateCommandResponse records the OBC response for a command.
func UpdateCommandResponse(ctx context.Context, commandID uuid.UUID, response string) error {
	err := Queries().UpdateCommandResponse(ctx, gen.UpdateCommandResponseParams{
		Response: &response,
		ID:       commandID,
	})
	if err != nil {
		return fmt.Errorf("failed to update response of command %v: %w", commandID, err)
	}
	return nil
}

// CreateTelemetry inserts one telemetry value into transactional.telemetry.
// timestamp is normalized to UTC before insertion.
func CreateTelemetry(ctx context.Context, telemetryType int32, value string, timestamp time.Time) error {
	err := Queries().CreateTelemetry(ctx, gen.CreateTelemetryParams{
		ID:        uuid.New(),
		Type:      telemetryType,
		Value:     &value,
		Timestamp: timestamp.UTC(),
	})
	if err != nil {
		return fmt.Errorf("failed to create telemetry of type %d: %w", telemetryType, err)
	}
	return nil
}

// ClearTelemetryByType deletes every telemetry row whose type_ matches
// telemetryType, and returns the number of rows deleted.
func ClearTelemetryByType(ctx context.Context, telemetryType int32) (int64, error) {
	deleted, err := Queries().ClearTelemetryByType(ctx, telemetryType)
	if err != nil {
		return 0, fmt.Errorf("failed to clear telemetry of type %d: %w", telemetryType, err)
	}
	return deleted, nil
}

// sessionStatus converts a SessionStatus into the value the sessionstatus
// Postgres enum expects.
func sessionStatus(status enums.SessionStatus) (gen.Sessionstatus, error) {
	if status == enums.SessionStatuses.EnumError {
		return "", fmt.Errorf("%w: unknown session status %d", ErrInvalidRow, status)
	}
	return gen.Sessionstatus(status.String()), nil
}

// commandStatus converts a CommandStatus into the value the commandstatus
// Postgres enum expects.
func commandStatus(status enums.CommandStatus) (gen.Commandstatus, error) {
	if status == enums.CommandStatuses.EnumError {
		return "", fmt.Errorf("%w: unknown command status %d", ErrInvalidRow, status)
	}
	return gen.Commandstatus(status.String()), nil
}

// validateMainCommand runs the data-integrity checks the ORM used to enforce on
// main.commands rows.
func validateMainCommand(command gen.MainCommand) error {
	switch {
	case command.DataSize < 0:
		return fmt.Errorf("%w: data_size must be >= 0", ErrInvalidRow)
	case command.TotalSize <= 0:
		return fmt.Errorf("%w: total_size must be > 0", ErrInvalidRow)
	case command.Priority < 0:
		return fmt.Errorf("%w: priority must be >= 0", ErrInvalidRow)
	}

	// params and format must either both be absent, or both list the same
	// number of comma-separated values.
	// TODO: check that the params have valid types
	switch {
	case command.Params == nil && command.Format == nil:
		return nil
	case command.Params == nil:
		return fmt.Errorf("%w: missing params", ErrInvalidRow)
	case command.Format == nil:
		return fmt.Errorf("%w: missing format", ErrInvalidRow)
	case strings.Count(*command.Params, ",") != strings.Count(*command.Format, ","):
		return fmt.Errorf("%w: params and format do not have the same number of values", ErrInvalidRow)
	}
	return nil
}
